c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,887 @@
+  auto stream = c10::cuda::getCurrentCUDAStream();
+  void* dummy;
+
+  // Defensively checks for preexisting CUDA error state.
+  auto err = cudaGetLastError();
+  C10_CUDA_CHECK(err);
+
+  while (true) {
+    // Duplicates some logic from mallocAsync to work with the error state
+    // directly instead of repeatedly catching an exception thrown by
+    // mallocAsync.
+    if (pytorch_used_bytes[device] + guess > pytorch_memory_limits[device]) {
+      err = cudaErrorMemoryAllocation;
+    } else {
+      err = cudaMallocAsync(&dummy, guess, stream);
+    }
+
+    if (err == cudaSuccess) {
+      cudaFreeAsync(dummy, stream);
+      *maxWorkspaceGuess = guess;
+      return;
+    } else if (err == cudaErrorMemoryAllocation) {
+      cudaGetLastError(); // clear CUDA error
+      guess >>= 1; // quick and dirty: try half the size next iteration
+    } else {
+      C10_CUDA_CHECK(err);
+    }
+  }
+}
+
+void* getBaseAllocation(void* ptr, size_t* size) {
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  auto it = ptr_info.find(ptr);
+  TORCH_INTERNAL_ASSERT(it != ptr_info.end(), "ptr not found in ptr_info");
+
+  if (size) {
+    *size = it->second.size;
+  }
+
+  return ptr;
+}
+
+void recordStream(const DataPtr& ptr, cuda::CUDAStream stream) {
+  std::lock_guard<std::mutex> lk(general_mutex);
+  auto ptr_val = ptr.get();
+  // Empty tensor's storage().data() might be a null ptr. As there is no
+  // blocks associated with those tensors, it is fine to do nothing here.
+  if (!ptr_val) {
+    return;
+  }
+
+  // The pointer should exist in the map already.
+  auto it = ptr_info.find(ptr_val);
+  TORCH_INTERNAL_ASSERT(it != ptr_info.end(), "ptr not found in ptr_info");
+
+  UsageStream to_record{stream.stream(), stream.device_index()};
+  if (to_record == it->second.creation_stream) {
+    TORCH_WARN(
+        "Called record_stream on tensor whose original creation stream "
+        "matches the recorded stream. This is unnecessary and has no effect.");
+  } else {
+    it->second.recorded_streams.insert(to_record);
+  }
+}
+
+std::mutex* getFreeMutex() {
+  return &cuda_free_mutex;
+}
+
+std::shared_ptr<void> getIpcDevPtr(std::string handle) {
+  TORCH_CHECK(
+      false,
+      "cudaMallocAsync does not yet support getIpcDevPtr. "
+      "If you need it, please file an issue describing your use case.");
+}
+
+// Collects stats for device.
+// If device hasn't been used yet, returns 0s without creating a context.
+DeviceStats getDeviceStats(int device) {
+  assertValidDevice(device);
+
+  // Memory currently reserved by the mempool
+  uint64_t reserved_mem_current = 0;
+  // High-water mark of memory reserved by the mempool since last reset
+  uint64_t reserved_mem_peak = 0;
+  // Memory currently in use by the mempool
+  uint64_t used_mem_current = 0;
+  // High-water mark of memory
+  uint64_t used_mem_peak = 0;
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  if (devs_initialized_flags[device]) {
```

Comment:
this complicates using once_flag for initializing devices. probably have to do our own fancy double-checked locking